documentation/modules/managing/con-consumer-config-properties.adoc

```diff
@@ -0,0 +1,252 @@
+
+----
+# ...
+fetch.max.bytes=52428800 <1>
+max.partition.fetch.bytes=1048576 <2>
+# ...
+----
+<1> The maximum amount of data in bytes returned for a fetch request.
+<2> The maximum amount of data in bytes returned for each partition.
+
+== Avoiding data loss or duplication when committing offsets
+
+The Kafka _auto-commit mechanism_ allows a consumer to commit the offsets of messages automatically.
+If enabled, the consumer will commit offsets received from polling the broker at 5000ms intervals.
+
+The auto-commit mechanism is convenient, but it introduces a risk of data loss and duplication.
+If a consumer has fetched and transformed a number of messages, but the system crashes with processed messages in the consumer buffer when performing an auto-commit, that data is lost.
+If the system crashes after processing the messages, but before performing the auto-commit, the data is duplicated on another consumer instance after rebalancing.
+
+To minimize the likelihood of data loss or duplication, set `enable.auto.commit` to `false` and develop your client application to have more control over committing offsets.
+Or use `auto.commit.interval.ms=1` to decrease the intervals between commits.
+
+By setting to `enable.auto.commit` to `false`, you can commit offsets after *all* processing has been performed and the message has been consumed.
+For example, you can set up your application to call the Kafka `commitSync` and `commitAsync` commit APIs.
+
+The `commitSync` API commits the offsets in a message batch returned from polling.
+You call the API when you are finished processing all the messages in the batch.
+If you use the `commitSync` API, the application will not poll for new messages until the last offset in the batch is committed.
+If this impacts throughput unduly, you can commit less frequently,
+or you can use the `commitAsync` API.
+The `commitAsync` API does not wait for the broker to respond to a commit request,
+but risks creating more duplicates when rebalancing.
+A common approach is to combine both commit APIs in an application, with the `commitSync` API used just before shutting the consumer down or rebalancing to make sure the final commit is successful.
+
+Also, consider using transactional ids and having idempotence enabled (`enable.idempotence=true`) on the producer side to guarantee exactly once delivery.
+On the consumer side, you can then use the `isolation.level` property to control how transactional messages are read by the consumer.
+
+The `isolation.level` property has two valid values:
+
+* `read_committed`
+* `read_uncommitted` (default)
+
+Use `read_committed` to ensure that only transactional messages that have been committed are read by the consumer.
+However, this will cause an increase in end-to-end latency, because the consumer won’t be able to return a message until the brokers have written the transaction markers that record the result of the transaction (_committed_ or _aborted_).
```

Comment:
```suggestion
However, this will cause an increase in end-to-end latency, because the consumer will not be able to return a message until the brokers have written the transaction markers that record the result of the transaction (_committed_ or _aborted_).
```